### PR TITLE
Eliminate `tempfile.mkstemp`

### PR DESCRIPTION
**Pull request by harupy**
## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

We should not use `tempfile.mkstemp` unless we really need it because it doesn't clean up the directory.

## How is this patch tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests (describe details, including test results, below)

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: tests/spark/autologging/datasource/test_spark_datasource_autologging_order.py [hunk 1]
```diff
@@ -28,8 +26,7 @@ def test_enabling_autologging_before_spark_session_works(disable):
     schema = StructType([StructField("number2", IntegerType())])
     rdd = spark_session.sparkContext.parallelize(rows)
     df = spark_session.createDataFrame(rdd, schema)
-    tempdir = tempfile.mkdtemp()
-    filepath = os.path.join(tempdir, "test-data")
+    filepath = os.path.join(tmp_path, "test-data")
```
Comment (harupy) [inline tests/spark/autologging/datasource/test_spark_datasource_autologging_order.py:29]: This can be simplified, but that's out of scope

**Review by BenWilson2 — APPROVED** [review]
LGTM

File: tests/tracking/test_tracking.py [hunk 2]
```diff
@@ -547,14 +546,12 @@ def test_log_artifact_with_dirs(tmpdir):
         ) == {os.path.basename(str(sub_dir))}
 
 
-def test_log_artifact():
-    artifact_src_dir = tempfile.mkdtemp()
+def test_log_artifact(tmp_path):
```
Comment (BenWilson2) [inline tests/tracking/test_tracking.py:549]: Do we just need to revert the changes in this test and manually clean up the temp directory? Looks like a sqlite instance got created which causes the diff to not match?

File: tests/tracking/test_rest_tracking.py [hunk 3]
```diff
@@ -751,7 +750,7 @@ def test_artifacts(mlflow_client, tmp_path):
     created_run = mlflow_client.create_run(experiment_id)
     assert created_run.info.artifact_uri.startswith(experiment_info.artifact_location)
     run_id = created_run.info.run_id
-    src_dir = tempfile.mkdtemp("test_artifacts_src")
+    src_dir = tmp_path.joinpath("test_artifacts_src")
```
Comment (BenWilson2) [inline tests/tracking/test_rest_tracking.py:753]:
```suggestion
    src_dir = tmp_path.joinpath("test_artifacts_src").mkdir(parents=True, exist_ok=True)
```
Looks like line 754-756 can be changed to:

```
src_file = src_dir.joinpath("my.file")
src_file.write_text("Hello, World!")
```

to make the tests pass

**Comment by BenWilson2** [comment]
This PR fixes the artifact example suite failure. If we can get those core python tests (and, by extension, that skinny test) to pass, we should be good! :D

File: tests/tracking/test_tracking.py (hunk 2 above)
Comment (harupy) [reply tests/tracking/test_tracking.py:549]: @BenWilson2 

Then, we can create a subdirectory in `tmp_path`, thanks for the catch!

File: tests/tracking/test_rest_tracking.py (hunk 3 above)
Comment (harupy) [reply tests/tracking/test_rest_tracking.py:753]: We can, but that's out of the scope of this PR :)

Comment (BenWilson2) [reply tests/tracking/test_rest_tracking.py:753]: AH that's right.... I totally forgot that os.path.join() can accept a pathlib.Path. That was why I suggested that change ;) It's not needed.

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:753]: wait, did this test fail?

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:753]: it did

Comment (harupy) [reply tests/tracking/test_rest_tracking.py:753]: `mkdir` was missing, thanks for the catch!
